### PR TITLE
Update Desktop Virtualization API to 2022-02-10; scheduled_agent_updates  supports scheduled_agent_updates

### PR DESCRIPTION
**Pull request by amccullough84**
Fixes https://github.com/hashicorp/terraform-provider-azurerm/issues/16542

### CODE REVIEW

**Comment by amccullough84** [comment]
Tests:

$ TF_ACC=1 go test -v ./internal/services/desktopvirtualization -run=TestAccVirtualDesktop -timeout 30m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccVirtualDesktopApplicationGroup_basic
=== PAUSE TestAccVirtualDesktopApplicationGroup_basic
=== RUN   TestAccVirtualDesktopApplicationGroup_complete
=== PAUSE TestAccVirtualDesktopApplicationGroup_complete
=== RUN   TestAccVirtualDesktopApplicationGroup_update
=== PAUSE TestAccVirtualDesktopApplicationGroup_update
=== RUN   TestAccVirtualDesktopApplicationGroup_requiresImport
=== PAUSE TestAccVirtualDesktopApplicationGroup_requiresImport
=== RUN   TestAccVirtualDesktopApplication_basic
=== PAUSE TestAccVirtualDesktopApplication_basic
=== RUN   TestAccVirtualDesktopApplication_complete
=== PAUSE TestAccVirtualDesktopApplication_complete
=== RUN   TestAccVirtualDesktopApplication_update
=== PAUSE TestAccVirtualDesktopApplication_update
=== RUN   TestAccVirtualDesktopApplication_requiresImport
=== PAUSE TestAccVirtualDesktopApplication_requiresImport
=== RUN   TestAccVirtualDesktopHostPoolRegInfo_basic
=== PAUSE TestAccVirtualDesktopHostPoolRegInfo_basic
=== RUN   TestAccVirtualDesktopHostPoolRegInfo_update
=== PAUSE TestAccVirtualDesktopHostPoolRegInfo_update
=== RUN   TestAccVirtualDesktopHostPool_basic
=== PAUSE TestAccVirtualDesktopHostPool_basic
=== RUN   TestAccVirtualDesktopHostPool_agentupdates
=== PAUSE TestAccVirtualDesktopHostPool_agentupdates
=== RUN   TestAccVirtualDesktopHostPool_complete
=== PAUSE TestAccVirtualDesktopHostPool_complete
=== RUN   TestAccVirtualDesktopHostPool_update
=== PAUSE TestAccVirtualDesktopHostPool_update
=== RUN   TestAccVirtualDesktopHostPool_requiresImport
=== PAUSE TestAccVirtualDesktopHostPool_requiresImport
=== RUN   TestAccVirtualDesktopScalingPlan_basic
=== PAUSE TestAccVirtualDesktopScalingPlan_basic
=== RUN   TestAccVirtualDesktopScalingPlan_complete
=== PAUSE TestAccVirtualDesktopScalingPlan_complete
=== RUN   TestAccVirtualDesktopScalingPlan_update
=== PAUSE TestAccVirtualDesktopScalingPlan_update
=== RUN   TestAccVirtualDesktopScalingPlan_requiresImport
=== PAUSE TestAccVirtualDesktopScalingPlan_requiresImport
=== RUN   TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_basic
=== PAUSE TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_basic
=== RUN   TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_complete
=== PAUSE TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_complete
=== RUN   TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_update
=== PAUSE TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_update
=== RUN   TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_requiresImport
=== PAUSE TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_requiresImport
=== CONT  TestAccVirtualDesktopApplicationGroup_basic
=== CONT  TestAccVirtualDesktopHostPool_complete
=== CONT  TestAccVirtualDesktopScalingPlan_requiresImport
=== CONT  TestAccVirtualDesktopApplication_update
--- PASS: TestAccVirtualDesktopHostPool_complete (103.88s)
=== CONT  TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_requiresImport
--- PASS: TestAccVirtualDesktopApplicationGroup_basic (105.18s)
=== CONT  TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_update
--- PASS: TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_requiresImport (123.12s)
=== CONT  TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_complete
--- PASS: TestAccVirtualDesktopScalingPlan_requiresImport (260.46s)
=== CONT  TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_basic
--- PASS: TestAccVirtualDesktopApplication_update (290.78s)
=== CONT  TestAccVirtualDesktopApplicationGroup_requiresImport
--- PASS: TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_complete (98.04s)
=== CONT  TestAccVirtualDesktopApplication_complete
--- PASS: TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_update (219.90s)
=== CONT  TestAccVirtualDesktopApplication_basic
--- PASS: TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_basic (97.24s)
=== CONT  TestAccVirtualDesktopApplicationGroup_update
--- PASS: TestAccVirtualDesktopApplicationGroup_requiresImport (113.43s)
=== CONT  TestAccVirtualDesktopApplicationGroup_complete
--- PASS: TestAccVirtualDesktopApplication_basic (84.95s)
=== CONT  TestAccVirtualDesktopHostPoolRegInfo_update
--- PASS: TestAccVirtualDesktopApplication_complete (88.69s)
=== CONT  TestAccVirtualDesktopHostPool_agentupdates
=== CONT  TestAccVirtualDesktopHostPool_basic
--- PASS: TestAccVirtualDesktopApplicationGroup_complete (86.15s)
=== CONT  TestAccVirtualDesktopHostPoolRegInfo_basic
--- PASS: TestAccVirtualDesktopApplicationGroup_update (197.36s)
--- PASS: TestAccVirtualDesktopHostPool_basic (81.03s)
=== CONT  TestAccVirtualDesktopApplication_requiresImport
--- PASS: TestAccVirtualDesktopHostPoolRegInfo_update (193.98s)
=== CONT  TestAccVirtualDesktopScalingPlan_basic
--- PASS: TestAccVirtualDesktopHostPool_agentupdates (193.59s)
=== CONT  TestAccVirtualDesktopScalingPlan_update
--- PASS: TestAccVirtualDesktopHostPoolRegInfo_basic (88.51s)
=== CONT  TestAccVirtualDesktopScalingPlan_complete
--- PASS: TestAccVirtualDesktopApplication_requiresImport (121.38s)
=== CONT  TestAccVirtualDesktopHostPool_requiresImport
--- PASS: TestAccVirtualDesktopScalingPlan_basic (197.01s)
=== CONT  TestAccVirtualDesktopHostPool_update
--- PASS: TestAccVirtualDesktopHostPool_requiresImport (111.13s)
--- PASS: TestAccVirtualDesktopHostPool_update (223.74s)
--- PASS: TestAccVirtualDesktopScalingPlan_update (496.10s)
--- PASS: TestAccVirtualDesktopScalingPlan_complete (635.84s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/desktopvirtualization 1279.542s

File: internal/services/desktopvirtualization/virtual_desktop_host_pool_resource.go [hunk 1]
```diff
@@ -139,6 +139,53 @@ func resourceVirtualDesktopHostPool() *pluginsdk.Resource {
...
+						},
+
+						"hour_of_day": {
+							Type:         pluginsdk.TypeInt,
+							Required:     true,
+							ValidateFunc: validation.IntBetween(0, 23),
+						},
+					},
+				},
+			},
```
Comment (katbyte) [inline internal/services/desktopvirtualization/virtual_desktop_host_pool_resource.go:187]: could we put this all in a block to clean up the schema? it also will match the API better

File: internal/services/desktopvirtualization/virtual_desktop_host_pool_resource.go [hunk 2]
```diff
@@ -238,6 +306,35 @@ func resourceVirtualDesktopHostPoolUpdate(d *pluginsdk.ResourceData, meta interf
...
+				updatesEnabled = d.Get("scheduled_agent_updates_enabled").(bool)
+				updatesScheduled := hostpool.SessionHostComponentUpdateTypeScheduled
+				updatesDefault := hostpool.SessionHostComponentUpdateTypeDefault
+				if updatesEnabled {
+					payload.Properties.AgentUpdate.Type = &updatesScheduled
+				} else {
+					payload.Properties.AgentUpdate.Type = &updatesDefault
+					payload.Properties.AgentUpdate.MaintenanceWindows = &[]hostpool.MaintenanceWindowPatchProperties{}
+				}
+			}
```
Comment (katbyte) [inline internal/services/desktopvirtualization/virtual_desktop_host_pool_resource.go:324]: could we do inlight assingments
```suggestion

				updatesScheduled := hostpool.SessionHostComponentUpdateTypeScheduled
				updatesDefault := hostpool.SessionHostComponentUpdateTypeDefault
				if d.Get("scheduled_agent_updates_enabled").(bool) {
					payload.Properties.AgentUpdate.Type = pointer.FromString(hostpool.SessionHostComponentUpdateTypeScheduled)
				} else {
					payload.Properties.AgentUpdate.Type = pointer.fromString()
					payload.Properties.AgentUpdate.MaintenanceWindows = &[]hostpool.MaintenanceWindowPatchProperties{}
				}
			}
```

File: website/docs/r/virtual_desktop_host_pool.html.markdown [hunk 3]
```diff
@@ -70,14 +76,33 @@ The following arguments are supported:
...
+* `agent_updates_schedule` - (Optional) An `agent_updates_schedule` block as defined below. A maximum of two blocks can be added. 
+
 * `tags` - (Optional) A mapping of tags to assign to the resource.
 
+---
+
+An `agent_updates_schedule` block supports the following:
+
+* `day_of_week` - (Required) The day of the week on which agent updates should be performed. Possible values are `Monday`, `Tuesday`, `Wednesday`, `Thursday`, `Friday`, `Saturday`, and `Sunday`
+* `hour_of_day` - (Required) The hour of day the update window should start. The update is a 2 hour period following the hour provided. The value should be provided as a number between 0 and 23, with 0 being midnight and 23 being 11pm. A leading zero should not be used.
```
Comment (katbyte) [inline website/docs/r/virtual_desktop_host_pool.html.markdown:102]: suggest this schema:
```
scheduled_agent_updates {
    enabled = false
    timezone = UTC
    use_session_host_timezone = false
    schedule {
        day_of_week
        hour_of_day
    }
}
```

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks for the pr @amccullough84 - I think we should put these properties in a block to clean up the resulting tf/schema and better matcht he api. i've let an example inline, WDYT?

File: website/docs/r/virtual_desktop_host_pool.html.markdown (hunk 3 above)
Comment (katbyte) [reply website/docs/r/virtual_desktop_host_pool.html.markdown:102]: looks like the docs need updating still?

Comment (amccullough84) [reply website/docs/r/virtual_desktop_host_pool.html.markdown:102]: yup, sorry - haven't had time, was waiting to reply until I had all the comments addressed. Should have this sorted by later today!

**Comment by amccullough84** [comment]
Hi @katbyte, thanks for the review and the comments. I've updated as requested, so is hopefully good now if you could take another look when you have a moment.

**Comment by amccullough84** [comment]
Hi @katbyte - just checking if there's anything outstanding on this PR?

**Review by katbyte — APPROVED** [review]
Thanks @amccullough84 - LGTM and running tests now 🖥️

**Pull request merged**